gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StorageStubProvider.java

```diff
@@ -153,12 +168,51 @@ private ChannelAndRequestCounter buildManagedChannel() {
     return new ChannelAndRequestCounter(channel, counter);
   }
 
-  public StorageBlockingStub getBlockingStub() {
-    return StorageGrpc.newBlockingStub(getManagedChannel());
+  public StorageBlockingStub getBlockingStub() throws IOException {
+    StorageBlockingStub blockingStub = StorageGrpc.newBlockingStub(getManagedChannel());
+    try {
+      if (!credentialFactory.useMetadataService()) {
+        GoogleCredential googleCredential = (GoogleCredential) credential;
+        GoogleCredentials credentials =
+            ServiceAccountCredentials.newBuilder()
+                .setPrivateKey(googleCredential.getServiceAccountPrivateKey())
+                .setPrivateKeyId(googleCredential.getServiceAccountPrivateKeyId())
+                .setClientEmail(googleCredential.getServiceAccountId())
+                .setScopes(googleCredential.getServiceAccountScopes())
+                .setTokenServerUri(new URI(googleCredential.getTokenServerEncodedUrl()))
+                .build();
+
+        blockingStub = blockingStub.withCallCredentials(MoreCallCredentials.from(credentials));
+      }
+    } catch (URISyntaxException e) {
+      throw new IOException("Token server URI could not be parsed.", e);
+    }
+    return blockingStub;
   }
 
-  public StorageStub getAsyncStub() {
-    return StorageGrpc.newStub(getManagedChannel()).withExecutor(backgroundTasksThreadPool);
+  public StorageStub getAsyncStub() throws IOException {
+    StorageStub asyncStub = StorageGrpc.newStub(getManagedChannel());
```

Comment:
I think it will be more readable and efficient if we wrap this in `if` and immediately return from it:

```java
if (credentialFactory.useMetadataService()) {
  return StorageGrpc.newBlockingStub(getManagedChannel());
}
```

Reply:
Actually, can we avoid this, and initialize credential channel from passed credential object always?

Reply:
I think branch is inevitable unless we can make the Credential used in default matadata service case a GoogleCredential. So the current logic would essentially be use GCE default SA (this will be done by grpc lib if we don't provide a credential on the stub) by default otherwise use the provided credential. I feel like this way is the most feasible and straightforward way to solve this problem without overengineering it.